projects/baby_driver/scripts/spi_exchange.py

```diff
@@ -0,0 +1,115 @@
+""" This modules provides the Python implementation to exchange data through the SPI protocol"""
+
+import can_util
+from gpio_port import GpioPort
+from message_defs import BabydriverMessageId
+
+GPIO_PINS_PER_PORT = 16
+
+# pylint: disable=invalid-name
+# pylint: disable=too-many-arguments
+# pylint: disable=too-many-branches
+# pylint: disable=too-many-locals
+def spi_exchange(tx_bytes, rx_len, spi_port=1, spi_mode=0, baudrate=6000000, cs=None):
+    """
+    Sends data through SPI
+        The tx_bytes is an int array
+        The rx_len is a non-negative int.
+        The spi_port is an int
+        The spi_mode is a non-negative int.
+        The baudrate is a non-negative int.
+        The CS pin, if not None, is a tuple (port, pin), where pin is an
+            int in [0, NUM_PINS_PER_PORT) and port is either an int in [0, NUM_GPIO_PORTS),
+            or a string 'a' through 'f', case insensitive
+    Args:
+        tx_bytes: The bytes to TX.
+        rx_len: Number of bytes to RX
+        spi_port: port of the pin to perform SPI exchange on (1 or 2)
+        spi_mode: SPI mode to use (0, 1, 2 or 3)
+        baudrate: baudrate to use
+        cs: chip select port and pin to use (defaults to (1, 1))
+    Raises:
+        AttributeError: if cs_port parameter is entered as a string and is invalid
+        ValueError: if spi_port, spi_mode, rx_len, cs_port, cs_pin don't meet their requirements
+        Exception: if a non-zero status code was received.
+    """
+
+    if spi_port not in (1, 2):
+        raise ValueError("ERROR: Expected SPI port 1 or 2")
+    if spi_mode not in (0, 1, 2, 3):
+        raise ValueError("ERROR: Expected mode between 0 and 3")
+    if not 0 <= len(tx_bytes) <= 256:
+        raise ValueError("ERROR: numbers of tx_bytes must be between 0 and 256")
+    if not 0 <= rx_len <= 256:
+        raise ValueError("ERROR: rx_len must be a non-negative integer between 0 and 256")
```

Comment:
The maximum value is actually 255 not 256 since 256 won't fit in a uint8. So the conditions and error messages should exclude 256. The tests are passing because a value >= 256 causes a ValueError later in can_pack.

Also, it's kind of ambiguous whether the `not` applies to the 0 or the whole `0 <= len(tx_bytes) <= 256`, please put brackets around the `0 <= len(tx_bytes) <= 256` and similarly with `rx_len`.